ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE TypeFamilies #-}
+
+module Ouroboros.Consensus.Mempool.API (
+    Mempool(..)
+  , ApplyTx(..)
+    -- * Re-exports
+  , PrevApplied(..)
+  ) where
+
+import           Control.Monad.Except
+import           Data.Sequence (Seq)
+
+import           Control.Monad.Class.MonadSTM
+
+import           Ouroboros.Consensus.Ledger.Abstract
+
+-- TODO: Maybe we should move 'PrevApplied' someplace else
+import           Ouroboros.Storage.LedgerDB.Conf (PrevApplied (..))
+
+class UpdateLedger b => ApplyTx b where
+  -- | Generalized transaction
+  --
+  -- The mempool (and, accordingly, blocks) consist of "generalized
+  -- transactions"; this could be "proper" transactions (transferring funds) but
+  -- also other kinds of things such as update proposals, delegations, etc.
+  type family GenTx b :: *
+
+  -- | Update the ledger state with a generalized transaction
+  --
+  -- If the transaction was previously applied, expensive checks that have
+  -- previously been confirmed (such as checking cryptographic signatures)
+  -- should be omitted.
+  applyTx :: PrevApplied
+          -> GenTx b
+          -> LedgerState b
+          -> Except (LedgerError b) (LedgerState b)
+
+-- | Mempool
+--
+-- The mempool is the set of transactions that should be included in the next
+-- block. In principle this is a /set/ of all the transactions that we receive
+-- from our peers. In order to avoid flooding the network with invalid
+-- transactions,  however, we only want to keep /valid/ transactions in the
+-- mempool. That raises the question: valid with respect to which ledger state?
+--
+-- We opt for a very simple answer to this: the mempool will be interpreted
+-- as a /list/ of transactions; which are validated strictly in order, starting
+-- from the current ledger state. This has a number of advantages:
+--
+-- * It's simple to implement and it's efficient. In particular, no search for
+--   a valid subset is ever required.
+-- * When producing a block, we can simply take the longest possible prefix
+--   of transactions that fits in a block.
+-- * It supports wallets that submit dependent transactions (where later
+--   transaction depends on outputs from earlier ones).
+data Mempool m blk = Mempool {
+      -- | Add a bunch of transactions (oldest to newest)
+      --
+      -- As long as we keep the mempool entirely in-memory this could live in
+      -- @STM m@; we keep it in @m@ instead to leave open the possibility of
+      -- persistence.
+      --
+      -- Returns the transactions that were not added because they could not
+      -- be validated (in undefined order).
```

Comment:
We could state this more strongly. They have been rejected by the mempool given the current ledger state. This should be considered a permanent error by the submitting code (though possibly not by the original submitting wallet). It means the transaction is either unconditionally invalid, or that it is at least invalid for the current chain and preceeding transactions in the mempool (and in practice these conditions are unlikely to change).

Reply:
Have reworded it as follows:

```
      -- The transactions will be validated, /in order/, before they are added
      -- to the mempool. Invalid transactions will be returned along with the
      -- validation error. In principle it is possible that such validation
      -- errors are transient; for example, it is possible that a transaction is
      -- rejected because one of its inputs is not /yet/ available in the UTxO
      -- (the transaction it depends on is not yet in the chain, nor in the
      -- mempool). In practice however it is likely that rejected transactions
      -- will still be rejected later, and should just be dropped.
      --
      -- It is important to note one important special case of transactions 
      -- being "invalid": a transaction will /also/ be considered invalid if
      -- /that very same transaction/ is already included on the blockchain
      -- (after all, by definition that must mean its inputs have been used).
      -- Rejected transactions are therefore not necessarily a sign of 
      -- malicious behaviour. Indeed, we would expect /most/ transactions that
      -- are reported as invalid by 'addTxs' to be invalid precisely because
      -- they have already been included. (Distinguishing between these two 
      -- cases can be done in theory, but it is expensive unless we have an
      -- index of transaction hashes that have been included on the blockchain.)
```

Seem OK?